src/extensions/idle.rs

```diff
@@ -100,37 +130,74 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
     /// Internal helper that doesn't consume self.
     ///
     /// This is necessary so that we can keep using the inner `Session` in `wait_keepalive`.
-    fn wait_inner(&mut self, reconnect: bool) -> Result<WaitOutcome> {
+    fn wait_inner<F>(&mut self, reconnect: bool, mut callback: F) -> Result<WaitOutcome>
+    where
+        F: FnMut(UnsolicitedResponse) -> bool,
+    {
         let mut v = Vec::new();
-        loop {
-            let result = match self.session.readline(&mut v).map(|_| ()) {
+        let result = loop {
+            let rest = match self.session.readline(&mut v) {
                 Err(Error::Io(ref e))
                     if e.kind() == io::ErrorKind::TimedOut
                         || e.kind() == io::ErrorKind::WouldBlock =>
                 {
-                    if reconnect {
-                        self.terminate()?;
-                        self.init()?;
-                        return self.wait_inner(reconnect);
+                    break Ok(WaitOutcome::TimedOut);
+                }
+                Ok(_len) => {
+                    //  Handle Dovecot's imap_idle_notify_interval message
+                    if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+                        v.clear();
+                        continue;
+                    }
+                    match parse_idle(&v) {
+                        (_rest, Some(Err(r))) => break Err(r),
+                        (rest, Some(Ok(response))) => {
+                            if !callback(response) {
+                                break Ok(WaitOutcome::MailboxChanged);
+                            }
+                            rest
+                        }
+                        (rest, None) => rest,
                     }
-                    Ok(WaitOutcome::TimedOut)
                 }
-                Ok(()) => Ok(WaitOutcome::MailboxChanged),
-                Err(r) => Err(r),
-            }?;
+                Err(r) => break Err(r),
+            };
 
-            // Handle Dovecot's imap_idle_notify_interval message
-            if v.eq_ignore_ascii_case(b"* OK Still here\r\n") {
+            // Update remaining data with unparsed data if needed.
+            if rest.is_empty() {
                 v.clear();
             } else {
-                break Ok(result);
+                // Assert on partial parse in debug builds - we expect to always parse all
+                // or none of the input buffer. On release builds, we still do the right thing.
+                debug_assert!(
```

Comment:
Continuing the discussion from https://github.com/jonhoo/rust-imap/pull/186#discussion_r606870729.

Yes, this is what I meant!

Looking at this version of the change, I think I now understand better what bothers me about this. We're using `rest.len() != v.len()` as a way of checking "is this a complete parse", when we already have a good place to do this — the `None` case for `match parse_idle`! I suspect this could would be much nicer if we simply moved all the bits in the `if` here up into the appropriate match arms above.

Reply:
Ahh. Okay. The reason the parse / match is separate is because I usually like to separate buffer management from parse result handling. Inside the parse match arms you're usually holding a ref to the buffer you parsed out of, so sometimes it's a hassle to try to do buffer management right there.

I pushed a change that integrates the buffer management (with debug_assert on unexpected partial parse), since we can get away with it here. I also added some comments on the match arms to indicate what they're handling.

Reply:
Beautiful, yes, I think that reads _much_ better. I agree with you in general that it's nice to do the buffer management separately, but I think in this instance the code reads much more clearly this way.